symengine/symbol.cpp

```diff
@@ -35,23 +35,49 @@ RCP<const Symbol> Symbol::as_dummy() const
     return dummy(name_);
 }
 
+#ifdef WITH_SYMENGINE_THREAD_SAFE
+std::atomic<size_t> Dummy::count_ = 0;
+#else
 size_t Dummy::count_ = 0;
+#endif
 
-Dummy::Dummy() : Symbol("_Dummy_" + to_string(count_))
+Dummy::Dummy()
+    : Symbol(default_Dummy_prefix_
+             + to_string(
+#ifdef WITH_SYMENGINE_THREAD_SAFE
+                 Dummy::count_.fetch_add(1, std::memory_order_relaxed)
+#else
+                 count_
+#endif
+                     ))
 {
     SYMENGINE_ASSIGN_TYPEID()
-    count_ += 1;
+#ifdef WITH_SYMENGINE_THREAD_SAFE
+    // this is inefficient: we should not construct Dummy instances using
+    // Dummy() (since we need the thread-safely incremented value *at
+    // construction of Symbol*).
+    dummy_index = std::stoul(get_name().substr(default_Dummy_prefix_len_));
```

Comment:
I'm not terribly happy with this solution. The crux is that in the constructor:
```c++
Dummy::Dummy() : Symbol(some_name) 
{
    ...
```
we are passing a name to the Symbol class which (rightfully) does not carry and dummy index. So if we want to keep the name of dummy in sync with the index we're in a conundrum, this was the least invasive "fix" I could come up with. If someone knows of another alternative I'm all ears. A builder function works (that's pretty much what `inline RCP<const Dummy> dummy()` is, and there we do the efficient thing...